src/tools/data_analysis/tool.py

```diff
@@ -239,13 +214,228 @@ async def agenerate_and_run_code(self, query_str) -> dict:
+            "executed_code": executable_code
+        }
+        
+        return Response(response=full_response, metadata=response_metadata)
+    def _generate_scikit_code(self, query_str: str) -> str:
+        context = self._get_table_context()
+        scikit_response_str = self._llm.predict(
+            self._scikit_prompt,
+            df_str=context,
+            query_str=query_str,
+            instruction_str=self._instruction_str_scikit,
+        )
+
+        if self._verbose:
+            logging.info(f"> Scikit Code Instructions:\n{scikit_response_str}\n")
+            scikit_response_str = "```\n" + scikit_response_str + "\n```\n" if not scikit_response_str.startswith(
+                "```") else scikit_response_str
+            run_sync(cl.Message(content=f"Generated Scikit Instructions:\n\n{scikit_response_str}\n").send())
+
+        return scikit_response_str
+
+    async def _generate_scikit_code_async(self, query_str: str) -> str:
+        context = self._get_table_context()
+        scikit_response_str = await self._llm.apredict(
+            self._scikit_prompt,
+            df_str=context,
+            query_str=query_str,
+            instruction_str=self._instruction_str_scikit,
+        )
 
+        if self._verbose:
+            logging.info(f"> Scikit Code Instructions:\n{scikit_response_str}\n")
+            scikit_response_str = "```\n" + scikit_response_str + "\n```\n" if not scikit_response_str.startswith(
+                "```") else scikit_response_str
+            await cl.Message(content=f"Generated Scikit Instructions:\n\n{scikit_response_str}\n").send()
 
+        return scikit_response_str
+    
+    def _execute_scikit_code(self, code_str: str) -> str:
+        """Execute the generated Scikit-learn code."""
+        try:
+            run_sync(cl.Message(content=f"Generated Code:\n```python\n{code_str}\n```").send())
+
+            output_capture = io.StringIO()
+
+            # Import necessary modules within the execution context
+            exec_globals = {
+            "pd": pd,
+            "np": np,
+            "plt": plt,
+            "sns": sns,  # Add this line to include seaborn
+            "df": self._df
+            }
+            exec("""
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn as sns
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import StandardScaler
+from sklearn.svm import SVC
+from sklearn.linear_model import LinearRegression
+from sklearn.metrics import mean_squared_error, accuracy_score, classification_report, confusion_matrix
+        """, exec_globals)
+
+            with contextlib.redirect_stdout(output_capture):
+                exec(code_str, exec_globals)
+
+            execution_output = output_capture.getvalue()
+            output_capture.close()
+
+            run_sync(cl.Message(content=f"Execution Output:\n```\n{execution_output}\n```").send())
```

Comment:
This could be the reason why in your demo, the plot generated are displayed in the desktop instead of directly in the chainlit UI

Reply:
Fixed, change to save the figure and then send it to chainlit UI because show_plot() get error when there is multiple figure